### PR TITLE
feat(inputs.http_listener_v2): Add unix socket mode

### PR DESCRIPTION
**Pull request by bazko1**
## Summary
Following changes add two new options to `http_listener_v2` -> `network` default `tcp` can be `unix` which results in `service_address` interpreted as unix socket path that will be created and listened on by HTTP server. 
`socket_mode` - for editing socket file permission settings, by default `""` can be for example `644` only available for `network: "unix"`

Why:
I would like to have possibility of writing HTTP messages over unix socket to push messages to telegraf using libcurl based tools. This allows me to limit access to writing/reading messages with built in users/groups based mechanisms instead of trying to limit access via some firewall based options. More explanation can be found in linked issue.

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves https://github.com/influxdata/telegraf/issues/15730

## Considerations
For now I added single unit test for unix network case I could instead extend some of the existing ones to be something like:
```
var networkCases = [2]string{"tcp", "unix"}
func TestSomething(){
for _, net := range networkCases {
               t.Run(net, func(
                       t *testing.T,
               ) {  ... test scenario for network ... }
}
```

I was thinking about directly using Socket from socket_listener but its `listener` field is made private and I decided to not change that thus some of the code is basically copied from its implementation for example the SocketMode part.

### CODE REVIEW

**Comment by bazko1** [comment]
!signed-cla

File: plugins/inputs/http_listener_v2/sample.conf [hunk 1]
```diff
@@ -3,6 +3,17 @@
   ## Address and port to host HTTP listener on
   service_address = ":8080"
 
+  ## Network type to listen to default tcp
+  # if set to unix service_address will be interpreted as unix socket path
+  # network = "tcp"
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/sample.conf:8]: How about extending the `service_address` to be able to take an URL? E.g. `service_address = ":8080"` or `service_address = "tcp://127.0.0.1:8080"` or `service_address = "unix:///var/spool/mysocket"`? If no schema can be recognized we assume `tcp`... What do you think?

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go [hunk 2]
```diff
@@ -231,8 +234,10 @@ func TestWriteHTTP(t *testing.T) {
 	require.EqualValues(t, 204, resp.StatusCode)
 
 	acc.Wait(2)
-	hostTags := []string{"server02", "server03",
-		"server04", "server05", "server06"}
+	hostTags := []string{
+		"server02", "server03",
+		"server04", "server05", "server06",
+	}
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2_test.go:240]: Could you please revert the change? Even though I agree it looks better with your change, let's try to minimize modified lines to get a more clear view of what _actually_ changed.

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go [hunk 3]
```diff
@@ -429,8 +436,10 @@ func TestWriteHTTPGzippedData(t *testing.T) {
 	require.NoError(t, resp.Body.Close())
 	require.EqualValues(t, 204, resp.StatusCode)
 
-	hostTags := []string{"server02", "server03",
-		"server04", "server05", "server06"}
+	hostTags := []string{
+		"server02", "server03",
+		"server04", "server05", "server06",
+	}
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2_test.go:442]: Same as above.

**Review by srebhan — COMMENTED** [review]
Thanks for your contribution @bazko1! I do have some smaller comments in the code, please take a look!

File: plugins/inputs/http_listener_v2/sample.conf (hunk 1 above)
Comment (bazko1) [reply plugins/inputs/http_listener_v2/sample.conf:8]: hi Yes that could be done I guess this will also make the configuration reflect what is in the socket_listener so it makes sense.

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go (hunk 2 above)
Comment (bazko1) [reply plugins/inputs/http_listener_v2/http_listener_v2_test.go:240]: Yeah will revert. Maybe you could consider introducing `gofumpt` as main linter? :)

File: plugins/inputs/http_listener_v2/README.md [hunk 4]
```diff
@@ -36,8 +36,19 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
 ```toml @sample.conf
 # Generic HTTP write listener
 [[inputs.http_listener_v2]]
-  ## Address and port to host HTTP listener on
+  ## Address to host HTTP listener on
+  # can be prefixed by protocol tcp, or unix if not provided defaults to tcp
+  # if unix network type provided it should be followed by absolute path for unix socket
   service_address = ":8080"
+  # service_address = "tcp://:8443"
+  # service_address = "unix:///tmp/telegraf.sock"
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/README.md:44]: By convention we all 'single-hashed' comments carry "default values" so please really use double hashes for comments...
```suggestion
  ## Address to host HTTP listener on, optionally including the protocol (tcp or unix).
  ## "unix" addresses should contain the absolute path for the socket, e.g.
  ##   service_address = "unix:///tmp/telegraf.sock"
  service_address = "tcp://:8080"
```

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 5]
```diff
@@ -151,18 +158,68 @@ func (h *HTTPListenerV2) Init() error {
...
+	protoRegex := regexp.MustCompile(`\w://`)
+	if !protoRegex.MatchString(h.ServiceAddress) {
+		h.ServiceAddress = "tcp://" + h.ServiceAddress
+	}
+
+	var u *url.URL
+	// Parse and check the address
+	if runtime.GOOS == "windows" && strings.HasPrefix(h.ServiceAddress, "unix://") {
+		u = &url.URL{Scheme: "unix", Path: strings.TrimPrefix(h.ServiceAddress, "unix://")}
+	} else {
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:170]: Is this really necessary? Parsing a unix URL should also work on Windows, shouldn't it?

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 6]
```diff
@@ -151,18 +158,68 @@ func (h *HTTPListenerV2) Init() error {
...
 	if err != nil {
 		return err
 	}
 	h.tlsConf = tlsConf
 	h.listener = listener
-	h.Port = listener.Addr().(*net.TCPAddr).Port
+
+	if u.Scheme == "tcp" {
+		h.Port = listener.Addr().(*net.TCPAddr).Port
+	}
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:208]: Is the port used somewhere? If not, we should not set it and deprecate the config option!

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go [hunk 7]
```diff
@@ -724,6 +727,69 @@ func TestServerHeaders(t *testing.T) {
...
+			continue
+		}
+		acc := &testutil.Accumulator{}
+		require.NoError(t, listener.Start(acc))
+		require.Equal(t, c.expectedAddress, listener.listener.Addr().String())
+		require.Equal(t, c.expectedPort, listener.Port)
+		listener.Stop()
+	}
+}
+
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2_test.go:792]: This test is problematic in CI environments. Imagine another test using the same port, if they by chance run at the same time one of them will fail and we get flaky tests. I would remove the test here or solely stick to the URL parsing part without actually "starting" the listener...

**Review by srebhan — COMMENTED** [review]
Thanks @bazko1, the code looks really good. Just some more small comments...

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go (hunk 7 above)
Comment (bazko1) [reply plugins/inputs/http_listener_v2/http_listener_v2_test.go:792]: Yes I agree such test might be problematic when running in parallel with others, but afaik they are run synchronously and there are also other tests that actually, start listening on port. 
Anyway my logic should be tested enough without `Start` as it is located in `Init` function so I will remove the `Start` method call.

File: plugins/inputs/http_listener_v2/http_listener_v2.go (hunk 5 above)
Comment (bazko1) [reply plugins/inputs/http_listener_v2/http_listener_v2.go:170]: Yes it was problematic as sample address will look like `unix://c:/Users/bazko1/temp/telegraf.socket` the `url.Parse` starts to interpret the part after `c:` as port and it results in an parse error.

You can check that in `socket_listener` tests for windows are skipped with comment that the unixgram is not supported even though the unix socket is.
See: https://github.com/influxdata/telegraf/blob/master/plugins/inputs/socket_listener/socket_listener_test.go#L125-L128

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go (hunk 7 above)
Comment (bazko1) [reply plugins/inputs/http_listener_v2/http_listener_v2_test.go:792]: Well actually the `Init` already calls the `net.Listen` so that the port gets busy and we need to stop the internal `dial.Listener`. 

I think my changes are big enough especially with special windows case that having unit test for url parsing would be nice so please reconsider the idea of fully removing test, but if you insist I will remove.

File: plugins/inputs/http_listener_v2/http_listener_v2.go (hunk 6 above)
Comment (bazko1) [reply plugins/inputs/http_listener_v2/http_listener_v2.go]: Looks like it isn't lets deprecate! - [dfbaf83](https://github.com/influxdata/telegraf/pull/15764/commits/dfbaf831e5757738a736018226361a58ba297094)

File: plugins/inputs/http_listener_v2/README.md (hunk 4 above)
Comment (bazko1) [reply plugins/inputs/http_listener_v2/README.md]: thanks will follow the convention from now on - [a239d9e](https://github.com/influxdata/telegraf/pull/15764/commits/a239d9ed32e4ab2e9c55f7fa5e511d1e71e014db)

File: plugins/inputs/http_listener_v2/http_listener_v2.go (hunk 5 above)
Comment (srebhan) [reply plugins/inputs/http_listener_v2/http_listener_v2.go:170]: I think the correct way is to use `unix:///C:/Users/bazko1/...` (i.e. three slashes) according to https://github.com/golang/go/commit/844b625ebcc7101e09fb87828a0e71db942a2416...

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go (hunk 7 above)
Comment (srebhan) [reply plugins/inputs/http_listener_v2/http_listener_v2_test.go]: Well starting the listener in `Init()` is also wrong, this should be done in `Start()`... Anyhow, as you didn't introduce this, I don't blame you. ;-)

Regarding the tests: Yes, the tests in this file/package are executes sequentially, but all packages are executed in parallel (in the best case)... So if e.g. `outputs.http` opens a server on this port we run into trouble... So it is acceptable to specify no port or port zero as this will allow the OS to choose a free port but the above will definitively bite us. Been there suffered that.

In any case, what you do is to test if `url.Parse` works I think and I guess we should assume this works at least as long as this test will open a fixed port... So either change the test/code to not starting listen or remove the fixed ports.

File: plugins/inputs/http_listener_v2/http_listener_v2.go (hunk 5 above)
Comment (bazko1) [reply plugins/inputs/http_listener_v2/http_listener_v2.go:170]: Yes great I tested it and this works I will remove the windows case then.

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go (hunk 7 above)
Comment (bazko1) [reply plugins/inputs/http_listener_v2/http_listener_v2_test.go]: Okay lets make it right then I agree that the test can be actually removed especially when windows unix path works fine with url.Parse. 
I will move the `.Listen` part to `Start` then.

Comment (bazko1) [reply plugins/inputs/http_listener_v2/http_listener_v2_test.go]: Done in [45792b5](https://github.com/influxdata/telegraf/pull/15764/commits/45792b54d0196db88386fbf12227f446dea2109f)

**Review by srebhan — APPROVED** [review]
Thanks for the nice PR @bazko1!

**Review by DStrand1 — APPROVED** [review]
